Clarify confusion between kraus operators and collapse operators

a-corni opened the pull request:
Given that [qutip.mesolve](https://qutip.org/docs/4.0.2/modules/qutip/mesolve.html) uses collapse operators and not kraus operators, I propose to delete the trace of kraus operators in our code and documentation. 
- Modify the name of the variable `kraus_ops` to `local_collapse_ops` in `set_config`.
- Delete the L_0 operators from both notebook and code. Delete condition that first effective operator should be an identity and that sum of probability*operators is equal to identity (that's a condition on kraus operators, not on lindblad operators).
- Replace the link to Preskill's notes to Nielsen and Chuang's "Quantum Computation and Quantum Information" to justify $`L_i = p/4 \sigma_i`$ in depolarizing noise.
- Replace probabilities by rates in notebook and replace p(1-p)^(n-1) by gamma. Hopefully, the results (in the notebook at least) are not changed significantly.
- [x] Rename variables refering to probabilities to rates. Could we keep probabilities and define the rate as gamma = proba / (duration of the sequence) @dehond ? 

Closes #608

dehond commented:
>     * [ ]  Rename variables refering to probabilities to rates. Could we keep probabilities and define the rate as gamma = proba / (duration of the sequence) @dehond ?

I'd suggest that ideally everything is relabeled as a rate so that nothing depends on the sequence length. If the decay rate of some excited state is 0.1/μs, the probability of decay during 1 μs is (approximately) 0.1, but how do we define the probability of decay during a 10 μs sequence, or longer?

If you'd like to turn it around we face similar issues: we know that the probability of decay during a 10 μs sequence is $1-e^{-1} = 0.63$, but it's not accurate to say that this means that the probability of decay during a 1 μs sequence is $0.63/10 = 0.063$.

a-corni commented:
> > ```
> > * [ ]  Rename variables refering to probabilities to rates. Could we keep probabilities and define the rate as gamma = proba / (duration of the sequence) @dehond ?
> > ```
> 
> I'd suggest that ideally everything is relabeled as a rate so that nothing depends on the sequence length. If the decay rate of some excited state is 0.1/μs, the probability of decay during 1 μs is (approximately) 0.1, but how do we define the probability of decay during a 10 μs sequence, or longer?
> 
> If you'd like to turn it around we face similar issues: we know that the probability of decay during a 10 μs sequence is 1−e−1=0.63, but it's not accurate to say that this means that the probability of decay during a 1 μs sequence is 0.63/10=0.063.

Thanks Julius, I was also puzzled by these issues. I will implement the replacement of variables with `_proba` into `_rate`. I will put a `DeprecationWarning` signaling their deprecation.

HGSilveri commented on pulser-core/pulser/backend/noise_model.py:
```diff
@@ -67,12 +67,12 @@ class NoiseModel:
-            operator.
-        eff_noise_opers: The operators for the effective noise model. The
-            first operator must be the identity.
+        dephasing_prob: The rate of a dephasing error occuring (in rad/µs).
+        depolarizing_prob: The rate (in rad/µs) at which a depolarizing
+            error occurs.
+        eff_noise_probs: The rate associated to each effective noise operator
+            (in rad/µs).
```
I agree with the switch from probabilities to rates but I don't think we should use the same variables for this. Instead, we should add new `*_rate` variables instead.
Then, the meaning of the `*_prob` variables can be kept but their default should be changed to `None` and the docstring should point out that the variable is deprecated and `*_rate` should be used instead. 
Additionally, if we detect that the value of `*_prob` is not `None` we raise a `DeprecationWarning` and say that the provided value will be converted to a rate.

a-corni replied:
Hey Henrique,
Yes that's exactly what I have in mind, just did not have time to implement it yet hehe

HGSilveri commented on tutorials/classical_simulation/Simulating with effective noise channels.ipynb:
```diff
@@ -15,25 +15,23 @@
     "\n",
```
<pre class="ql-syntax" spellcheck="false">Line #8.        print("Noise probability = ", noise_rate)
</pre><p>Missed this one</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pasqal-io/Pulser/pull/616/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

a-corni replied:
Yeah turns out I did not save my last modifications before pushing, there are lots of "prob" left...

HGSilveri commented on tutorials/classical_simulation/Simulating with effective noise channels.ipynb:
```diff
@@ -15,25 +15,23 @@
     "\n",
```
<pre class="ql-syntax" spellcheck="false">Line #13.            noise="depolarizing", depolarizing_prob=noise_rate
</pre><p>Idem here</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pasqal-io/Pulser/pull/616/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

HGSilveri replied:
<pre class="ql-syntax" spellcheck="false">Line #11.        config_dephasing = SimConfig(noise="dephasing", dephasing_prob=noise_rate)
</pre><p>Same everywhere on this cell</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pasqal-io/Pulser/pull/616/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='62'/>

HGSilveri commented on tutorials/classical_simulation/Simulating with effective noise channels.ipynb:
```diff
@@ -15,25 +15,23 @@
     "\n",
```
<pre class="ql-syntax" spellcheck="false">Line #3.    noise_rates = np.round(np.linspace(0, 0.2, 5), 3)
</pre><p>So I guess we don't have to be restricted from 0 to 0.2 anymore?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pasqal-io/Pulser/pull/616/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='62'/>

a-corni replied:
No we don't have to, but I don't think we would see much if going to higher rates.

a-corni commented on pulser-core/pulser/backend/noise_model.py:
```diff
@@ -84,12 +91,40 @@ class NoiseModel:
+                )
+                if prob != rate:
+                    warnings.warn(
+                        f"Setting {rate_name} with the value from "
+                        f"{prob_name}.",
+                        UserWarning,
+                    )
+                    self._change_attribute(rate_name, prob)
```
I was not a fan of checking whether or not the rate was at its default value, but combining raising an error and a warning makes sense. Would you raise an error if the person provides the same rate and prob ?

a-corni replied:
Ah now I remember there is a situation in our code where we end up providing a rate and prob to a NoiseModel: when we make a SimConfig and then convert it to a noise model (as is happening in simulation.py). So I am not going to raise an error if the user provides same rate and prob.

a-corni commented on pulser-core/pulser/backend/noise_model.py:
```diff
@@ -84,12 +91,40 @@ class NoiseModel:
+            rate_name = f"{noise}_rate{'s' if noise=='eff_noise' else ''}"
+            prob, rate = (getattr(self, prob_name), getattr(self, rate_name))
+            if len(prob) > 0 if noise == "eff_noise" else prob is not None:
+                warnings.warn(
+                    f"{prob_name} is deprecated. Use {rate_name} instead.",
+                    DeprecationWarning,
+                )
```
What do you mean by "normal" operation ? In a notebook ?

HGSilveri replied:
For example, can be a script too. In general, whenever users are using the code directly I think deprecation warnings don't show by default. I've had this happen before, at least.

a-corni commented on pulser-simulation/pulser_simulation/hamiltonian.py:
```diff
@@ -106,95 +105,38 @@ def config(self) -> NoiseModel:
-                raise NotImplementedError(
-                    "Cannot include general "
-                    + "noise in digital- or all-basis."
+            basis_check("general")
```
Yeah that's what we were doing... It would make more sense to display it at "effective" instead.

HGSilveri replied:
I don't know about "general", "arbitrary" or "custom" would make more sense to me. In any case, we are not doing the renaming here so we could defer the decision to a future PR. I'm just afraid that if you change to "general" here and then we decide to go with something else, this might be forgotten.

a-corni replied:
So we were already going for "General" in several places, but I find it does not make much sense... This is why I decided to go with "effective" instead

a-corni commented on tutorials/classical_simulation/Simulating with effective noise channels.ipynb:
```diff
@@ -15,25 +15,23 @@
     "\n",
```
<p>Ah that's weird I thought I had modified this... In the end I don't think I will speak about probabilities - because I think the formula $`p(t) = \frac{1 - e^{-\gamma t}}{2}$ is true for dephasing noise, but is not for depolarizing noise - I don't think we have a formula for the depolarizing probability function of time @dehond ?</p>

a-corni replied:
![image](https://github.com/pasqal-io/Pulser/assets/61453516/b8612c56-1a09-4ffa-88ba-6f09080c3cf8)

a-corni replied:
![image](https://github.com/pasqal-io/Pulser/assets/61453516/1131aeb0-f74a-4326-a9e4-989fc492d8f7)

dehond replied:
> I don't think we have a formula for the depolarizing probability function of time @dehond ?

I suppose one could describe depolarizing noise as a decay envelope towards some stationary value, but this is only true in some limiting case (the same applies to dephasing noise). The expression given here is the decay of the contrast envelope for a single-atom Ramsey experiment. When interactions are introduced in a system this becomes more complicated, and in general it's not possible to describe it with such a clean analytical expression.

a-corni commented:
> I think the notebook needs to be executed again

It should be good now :)

HGSilveri commented on pulser-core/pulser/backend/noise_model.py:
```diff
@@ -111,6 +111,16 @@ def __post_init__(self) -> None:
+                        len(rate) > 0
+                        if noise == "eff_noise"
+                        else rate
+                        != self.__dataclass_fields__[rate_name].default
```
I know it's more verbose but I think it would be preferable to search through `dataclasses.fields(self)` to get the `rate` field and then get the default value from it.

There's no guarantee that the structure of the internal won't change, so we shouldn't access it.

a-corni commented:
<p>It takes 28sec to run the whole notebook, that's eventually a bit too long indeed</p>

---

 <a href='https://app.reviewnb.com/pasqal-io/Pulser/pull/616/discussion/#comment-1855631186'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

HGSilveri reviewed (approved):
LGTM, nice job!

a-corni commented:
@dehond Do you have any suggestion about the notebook ? You can preview it [here](https://pulser--616.org.readthedocs.build/en/616/tutorials/kraus_ops.html)😄

dehond commented:
> @dehond Do you have any suggestion about the notebook ? You can preview it [here](https://pulser--616.org.readthedocs.build/en/616/tutorials/kraus_ops.html)😄

On first sight it looks pretty good to me. Unfortunately I don't have time to review the code now and use it a bit myself. I can do it after Christmas if you'd like, but I can imagine you don't want to wait with the merge until then.

One question that comes up: is it possible to set $\gamma > 1$ now? I don't think this was possible before since it was interpreted as a probability. If it's a rate this limitation shouldn't exist, and it may be interesting to show it as an example case in one of the plots in the notebook.

a-corni commented:
> > @dehond Do you have any suggestion about the notebook ? You can preview it [here](https://pulser--616.org.readthedocs.build/en/616/tutorials/kraus_ops.html)😄
> 
> On first sight it looks pretty good to me. Unfortunately I don't have time to review the code now and use it a bit myself. I can do it after Christmas if you'd like, but I can imagine you don't want to wait with the merge until then.
> 
> One question that comes up: is it possible to set γ>1 now? I don't think this was possible before since it was interpreted as a probability. If it's a rate this limitation shouldn't exist, and it may be interesting to show it as an example case in one of the plots in the notebook.

Hey Julius, it's now possible to define $`\gamma>1`$ ! I have just updated the rates in the notebook (+ shorten a bit the duration of the sequence, change the color of the theoretical curve and fixed a typo) such that the notebook shows the outcome for the rates 0, 0.5, 1.0 and 1.5.
Other than that, I think it's okay to merge the PR now, we can always improve it after Christmas :)

HGSilveri reviewed (approved):
The new plots are very nice, great job!
